File: FE/src/utils/strokeMatching.ts
export interface Point {
  x: number;
  y: number;
}

export interface StrokeTransform {
  scale: number;
  offsetX: number;
  offsetY: number;
}


export const parseSVGPath = (
  d: string,
  numPoints: number = 50,
  scaleX = 1,
  scaleY = 1,
  transform: StrokeTransform = { scale: 1, offsetX: 0, offsetY: 0 }
): Point[] => {
  const pathElem = document.createElementNS("http://www.w3.org/2000/svg", "path");
  pathElem.setAttribute("d", d);
  const totalLength = pathElem.getTotalLength();

  const points: Point[] = [];

  if (totalLength <= 0) return points;

  for (let i = 0; i < numPoints; i++) {
    const dist = (i / (numPoints - 1)) * totalLength;
    const p = pathElem.getPointAtLength(dist);
    const tx = (p.x * transform.scale + transform.offsetX) * scaleX;
    const ty = (p.y * transform.scale + transform.offsetY) * scaleY;
    points.push({ x: tx, y: ty });
  }
  return points;
};


export const resamplePath = (path: Point[], numPoints: number = 50): Point[] => {
  if (path.length === 0) return [];
  if (path.length === 1) return Array(numPoints).fill(path[0]);
  let totalLength = 0;
  const distances: number[] = [0];
  for (let i = 1; i < path.length; i++) {
    const dx = path[i].x - path[i - 1].x;
    const dy = path[i].y - path[i - 1].y;
    const dist = Math.sqrt(dx * dx + dy * dy);
    totalLength += dist;
    distances.push(totalLength);
  }


  if (totalLength === 0) return Array(numPoints).fill(path[0]);

  const resampled: Point[] = [];
  resampled.push(path[0]);

  const step = totalLength / (numPoints - 1);
  let currentDist = step;
  let currentIdx = 1;

  for (let i = 1; i < numPoints - 1; i++) {
    while (currentIdx < path.length && distances[currentIdx] < currentDist) {
      currentIdx++;
    }


    const p1 = path[currentIdx - 1];
    const p2 = path[currentIdx];
    const d1 = distances[currentIdx - 1];
    const d2 = distances[currentIdx];


    if (d1 === d2) {
      resampled.push({ x: p2.x, y: p2.y });
    } else {
      const t = (currentDist - d1) / (d2 - d1);
      const x = p1.x + (p2.x - p1.x) * t;
      const y = p1.y + (p2.y - p1.y) * t;
      resampled.push({ x, y });
    }
    currentDist += step;
  }

  resampled.push(path[path.length - 1]);
  return resampled;
};


export const calculateDistance = (path1: Point[], path2: Point[]): number => {
  if (path1.length !== path2.length) return Infinity;
  let sum = 0;
  for (let i = 0; i < path1.length; i++) {
    const dx = path1[i].x - path2[i].x;
    const dy = path1[i].y - path2[i].y;
    sum += Math.sqrt(dx * dx + dy * dy);
  }
  return sum / path1.length;
};


export const matchStroke = (
  userPath: Point[],
  standardSVGPathStr: string,
  canvasWidth: number,
  canvasHeight: number,
  transform: StrokeTransform = { scale: 1, offsetX: 0, offsetY: 0 }
) => {
  const scaleX = canvasWidth / 109;
  const scaleY = canvasHeight / 109;

  const numPoints = 30;
  const standardPoints = parseSVGPath(standardSVGPathStr, numPoints, scaleX, scaleY, transform);
  const userPointsResampled = resamplePath(userPath, numPoints);

  const distance = calculateDistance(userPointsResampled, standardPoints);


  const userPointsReversed = [...userPointsResampled].reverse();
  const reverseDistance = calculateDistance(userPointsReversed, standardPoints);
  const threshold = 55;

  const isCorrect = distance < threshold;
  const isReversed = reverseDistance < distance && reverseDistance < threshold;

  return {
    isCorrect,
    isReversed,
    distance,
    reverseDistance,
    standardPoints,
  };
};

File: mirai_mobile/lib/utils/stroke_matching.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:path_drawing/path_drawing.dart';

class StrokeMatchResult {
  final bool isCorrect;
  final bool isReversed;
  final double distance;
  final double reverseDistance;
  final List<Offset> standardPoints;

  StrokeMatchResult({
    required this.isCorrect,
    required this.isReversed,
    required this.distance,
    required this.reverseDistance,
    required this.standardPoints,
  });
}

class StrokeMatching {
  static List<Offset> resamplePath(List<Offset> path, {int numPoints = 30}) {
    if (path.isEmpty) return [];
    if (path.length == 1) return List.filled(numPoints, path[0]);
    
    double totalLength = 0;
    List<double> distances = [0];
    
    for (int i = 1; i < path.length; i++) {
      final dx = path[i].dx - path[i - 1].dx;
      final dy = path[i].dy - path[i - 1].dy;
      final dist = sqrt(dx * dx + dy * dy);
      totalLength += dist;
      distances.add(totalLength);
    }

    if (totalLength == 0) return List.filled(numPoints, path[0]);

    List<Offset> resampled = [path[0]];
    final step = totalLength / (numPoints - 1);
    double currentDist = step;
    int currentIdx = 1;

    for (int i = 1; i < numPoints - 1; i++) {
      while (currentIdx < path.length && distances[currentIdx] < currentDist) {
        currentIdx++;
      }

      final p1 = path[currentIdx - 1];
      final p2 = path[currentIdx];
      final d1 = distances[currentIdx - 1];
      final d2 = distances[currentIdx];

      if (d1 == d2) {
        resampled.add(p2);
      } else {
        final t = (currentDist - d1) / (d2 - d1);
        final x = p1.dx + (p2.dx - p1.dx) * t;
        final y = p1.dy + (p2.dy - p1.dy) * t;
        resampled.add(Offset(x, y));
      }
      currentDist += step;
    }

    resampled.add(path.last);
    // Safety check just in case float math goes quirky
    if (resampled.length > numPoints) resampled = resampled.take(numPoints).toList();
    if (resampled.length < numPoints) {
      while (resampled.length < numPoints) {
        resampled.add(path.last);
      }
    }
    return resampled;
  }

  static double calculateDistance(List<Offset> path1, List<Offset> path2) {
    if (path1.length != path2.length) return double.infinity;
    double sum = 0;
    for (int i = 0; i < path1.length; i++) {
      final dx = path1[i].dx - path2[i].dx;
      final dy = path1[i].dy - path2[i].dy;
      sum += sqrt(dx * dx + dy * dy);
    }
    return sum / path1.length;
  }

  static List<Offset> parseSvgPath(String svgPathStr, double scaleX, double scaleY, {int numPoints = 30}) {
    // Parse using path_drawing
    final Path dartPath = parseSvgPathData(svgPathStr);
    
    // Scale the path
    final Matrix4 matrix = Matrix4.identity()..scale(scaleX, scaleY);
    final Path scaledPath = dartPath.transform(matrix.storage);

    final List<Offset> points = [];
    final metrics = scaledPath.computeMetrics().toList();
    
    if (metrics.isEmpty) return [];

    // Assuming a single continuous contour for one stroke
    final metric = metrics.first; 
    final totalLength = metric.length;

    if (totalLength <= 0) return [];

    for (int i = 0; i < numPoints; i++) {
      final dist = (i / (numPoints - 1)) * totalLength;
      final tangent = metric.getTangentForOffset(dist);
      if (tangent != null) {
        points.add(tangent.position);
      }
    }
    
    return points;
  }

  static StrokeMatchResult matchStroke(
    List<Offset> userPath,
    String standardSVGPathStr,
    double canvasWidth,
    double canvasHeight,
  ) {
    // Original web app SVG was laid out on a 109x109 canvas viewport
    final scaleX = canvasWidth / 109.0;
    final scaleY = canvasHeight / 109.0;
    const numPoints = 30;

    final standardPoints = parseSvgPath(standardSVGPathStr, scaleX, scaleY, numPoints: numPoints);
    if (standardPoints.isEmpty) {
      return StrokeMatchResult(
        isCorrect: false,
        isReversed: false,
        distance: double.infinity,
        reverseDistance: double.infinity,
        standardPoints: [],
      );
    }

    final userPointsResampled = resamplePath(userPath, numPoints: numPoints);
    final distance = calculateDistance(userPointsResampled, standardPoints);

    final userPointsReversed = userPointsResampled.reversed.toList();
    final reverseDistance = calculateDistance(userPointsReversed, standardPoints);

    // The threshold from the web version
    const threshold = 55.0;

    final isCorrect = distance < threshold;
    final isReversed = reverseDistance < distance && reverseDistance < threshold;

    return StrokeMatchResult(
      isCorrect: isCorrect,
      isReversed: isReversed,
      distance: distance,
      reverseDistance: reverseDistance,
      standardPoints: standardPoints,
    );
  }
}
